docs/datatypes.md

```diff
@@ -0,0 +1,253 @@
+type inside:
+
+    (datatype int-list
+      (:empty)
+      (:node int int-list)
+      )
+
+Now we can create a list of ints:
+
+    (int-list:node 1 (int-list:node 2 (int-list:empty)))
+
+Recursive datatypes _must_ have at least one variant that's not
+recursive, otherwise only infinite values are possible.
+
+We can make a different version of `int-list` that disallows
+empty lists:
+
+    (datatype ints2
+      (:last int)
+      (:more int ints2)
+      )
+
+    (ints2:more 1 (ints2:more 2 (ints2:last 3)))
+
+We can also have mutually recursive datatypes:
+
+    (datatype crazy1
+      (:one crazy2)
+      (:two crazy2 crazy2)
+      )
+
+    (datatype crazy2
+      (:stop)
+      (:more crazy1 crazy1)
+      )
+
+    (crazy2:more
+     (crazy1:two (crazy2:stop) (crazy2:stop))
+     (crazy1:one (crazy2:stop)))
+
+Recursive Polymorphic Datatypes
+-------------------------------
+
+Now, witness the firepower of this fully operational type system.
+
+We can create a list of anything:
+
+    (datatype list
+      (:cons 'a (list 'a))
+      (:nil)
+      )
+
+[this is the actual definition used in the Irken library]
+
+    (list:cons 1 (list:cons 2 (list:cons 3 (list:nil))))
+
+Or a binary tree of anything:
+
+    (datatype btree
+       (:empty)
+       (:node 'a (btree 'a) (btree 'a))
+       )
+
+Here's a `(tree int)`:
+
+    (btree:node 19
+      (btree:node 34
+        (btree:empty)
+        (btree:empty))
+      (btree:empty))
+
+and a `(btree string)`:
+
+    (btree:node "hey there" (btree:empty) (btree:empty))
+
+and a `(btree (list int))`:
+
+    (btree:node (list:cons 1 (list:nil)) (btree:empty) (btree:empty))
+
+Complex Datatypes
+-----------------
+
+By combining recursion and polymorphism, we can build complex and useful new types:
+
+    (datatype alist
+      (:empty)
+      (:node 'a 'b (alist 'a 'b))
+      )
+
+This is similar to an `alist` (or 'association list') from Lisp.
+
+    (alist:node 3 "three"
+      (alist:node 4 "four"
+        (alist:empty)))
+
+You could use this datatype to search for the value of type `'b`
+associated with a key of type `'a`.
+
+A Datatype For Mathematical Expressions
+---------------------------------------
+
+We can represent mathematical expressions symbolically.
+
+First, we create a datatype of all the different operations we support:
+
+    (datatype binop
+      (:+)
+      (:-)
+      (:*)
+      (:/)
+      )
+
+[Note that none of its variants have data attached.  This would be the equivalent
+ of an `enum` type in C, Java, etc...]
+
+Now, a datatype for expressions:
+
+    (datatype exp
+      (:var symbol)
+      (:int int)
+      (:binop binop exp exp)
+      )
+
+Here's a variable (named with a 'symbol'):
+
+    (exp:var 'x)
+
+Here's an integer:
+
+    (exp:int 42)
+
+Here we have `x + 42`:
+
+    (exp:binop (binop:+) (exp:var 'x) (exp:int 42))
+
+and `2*x-12`:
+
+    (exp:binop
+       (binop:-)
+       (exp:binop (binop:*) (exp:int 2) (exp:var 'x))
+       (exp:int 12))
+
```

Comment:
hrm. it got kind of abstract there towards the end. I understand it all, but I'm not sure what point is being made. maybe I expand once I _really_ grok it.